File: mergesort/python/mergesort.py
#!/usr/bin/env python

# mergesort.py
# Avg.   O(n log(n))
# Worst: O(n log(n))
# Author: Erik Bergenholtz

from random import randint

def merge(left, right):
    merged = []
    while len(left) != 0 and len(right) != 0:
        if left[0] >= right[0]:
            merged.append(left[0])
            left = left[1:]
        else:
            merged.append(right[0])
            right = right[1:]
    for i in left:
        merged.append(i)
    for i in right:
        merged.append(i)
    return merged


def mergesort(arr):
    if len(arr) == 1:
        return arr
    return merge(mergesort(arr[:len(arr)/2]),mergesort(arr[len(arr)/2:]))

def main():
    arr = []
    while len(arr) < 100:
        arr.append(randint(0,100))
    print arr
    print mergesort(arr)

if __name__ == '__main__':
    main()

File: mergesort/haskell/mergesort.hs
-- mergesort.hs
-- Avg.   O(n log(n))
-- Worst: O(n log(n))
-- Author: Erik Bergenholtz

import System.IO
import System.Random
import Data.List

merge :: Ord a => [a] -> [a] -> [a]
merge [] [] = []
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys)
    | x>=y = x:(merge xs (y:ys))
    | otherwise = y:(merge (x:xs) ys)

mergesort :: Ord a => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort xs = merge (mergesort left) (mergesort right)
    where left = take ( length xs `div` 2 ) xs
          right = drop ( length xs `div` 2) xs

randomize :: Int -> StdGen -> [Int]
randomize n = take n . unfoldr (Just . randomR (0,100))

printarr [x] = do
    putStrLn x
printarr (x:xs) = do
    putStrLn x
    print xs

main = do
    seed <- newStdGen
    let rs = randomize 50 seed
    print $ mergesort rs
